Added bc for rel. ops, negation, unary minus

JanJecmen opened the pull request:
Plus some minor formatting and script to plot benchmark speedup of one given revision

JanJecmen committed:
Benchmark revision and plot speedup

JanJecmen committed:
typo fix

JanJecmen committed:
Rel ops added

JanJecmen committed:
Benchmark script disable bc for installing

JanJecmen committed:
Merge branch 'insns' into insns2

JanJecmen committed:
No bytecode fix

JanJecmen committed:
Merge branch 'insns2' into insns

JanJecmen committed:
relops refactor

JanJecmen committed:
Unary + and -

JanJecmen committed:
logical not operator

JanJecmen committed:
test tweaks

JanJecmen committed:
Merge remote-tracking branch 'upstream/master'

JanJecmen committed:
insns.h minor reformatting

JanJecmen committed:
Speedup benchmark script output change

Changed output of the script to allow easier sorting by commit date when plotting history.

o- commented on tools/bench-speedup.sh:
```diff
@@ -0,0 +1,130 @@
+  # Mac OSX
+    F77="gfortran -arch x86_64" FC="gfortran -arch x86_64" CXXFLAGS="-g3 -O2" CFLAGS="-g3 -O2" ./configure --enable-R-shlib --without-internal-tzcode --with-ICU=no --disable-byte-compiled-packages
+else
+    CXXFLAGS="-g3 -O2" CFLAGS="-g3 -O2" ./configure --with-ICU=no --disable-byte-compiled-packages
```
--disable-byte-compiled-packages will give you quite different perf

maybe it's a fairer comparison? but still, not comparable to what people will get out of the box

JanJecmen replied:
Yeah, the idea was to take the ast interpreter as baseline and then see how gnu-r compiler and rir matched against each other. I thought that it would be better if both had to compile everything just in time during the benchmark, so rir wouldn't use gnu-r bytecode from the default packages.

But it's true that out of the box users will get the base and recommended packages pre-compiled, and rir will run on top of that, so maybe this is unnecessary.

o- replied:
Actually running RIR with R_ENABLE_JIT should recompile the package code. As soon rir is loaded then bytecompiled code and uncompiled code will be treated the same. So I think disabling precompiling packages is unfair to the stock R. That's also why we do one round of `execute()` without measuring as warmup. Otherwise it would be unfair to us, since we compile all the package code at runtime.

o- replied:
(Sidenote, maybe we should switch to R_ENABLE_JIT=3 to be extra sure...)

JanJecmen replied:
Oh ok... shouldn't have assumed that rir just runs gnu-r bytecode when it sees it.
Thanks for the explanation!

o- commented on tools/bench-speedup.sh:
```diff
@@ -0,0 +1,130 @@
+# checkout rir revision and build rir
+git checkout $RIR_REVISION
+
+ninja clean || true
```
should probably use `cmake --build`. Can't assume everyone to use ninja

JanJecmen replied:
OK, good point. Probably should then also change https://github.com/reactorlabs/rir/blob/master/tools/benchmark.sh#L34 and https://github.com/reactorlabs/rir/blob/master/tools/benchmark-revisions.sh#L35?

o- replied:
oh I see, I am correcting my mistakes and not yours :) As you want

o- commented on rir/src/interpreter/interp.c:
```diff
@@ -1910,40 +1910,212 @@ INSTRUCTION(sub_) {
+    ostack_push(ctx, res);
+}
+
+INSTRUCTION(not_) {
```
wouldn't you expect booleans to be common here? I think they should be in the fast path here and maybe also for others like `==`

JanJecmen replied:
Thanks, will change that..

o- reviewed (commented):
Hi, awesome. Thanks for creating the patch. It looks good to me except for some inline comments. Do you measure speedups? One of the travis jobs timed out. Sometimes it indicates a performance regression, but sometimes it's just travis being slow...

JanJecmen committed:
Change ninja to generic cmake --build in benchmark scripts

Benchmarking scripts assumed ninja to build rir, change this to be more portable.

JanJecmen committed:
Fix speedup benchmarking script

Change so that it doesn't disable default package compilation when installing R.
Also add script to plot history of rir speed by itself.

JanJecmen committed:
Add fast paths for logical scalars

o- commented:
thanks!
